### ISSUE TITLE
Questionable behaviour of symmetric_difference_all

### ISSUE THREAD

**Issue by martinfleis** [shapely/shapely#2027]
My understanding of `symmetric_difference_all` is that it returns all parts of geometries that are unique to a single one, i.e. those parts that do not intersect any other. But that doesn't seem to  be the case as shown in the example below.

```py
import numpy as np
import shapely

geoms = np.array(
    [box(0, 0, 2, 2), box(1, 1, 3, 3), box(0, 0, 1.5, 1.5)]
)
shapely.symmetric_difference_all(geoms)
```

Visually:

You can see that the small square in the middle is shared by all three polygons.
![Screenshot 2024-03-25 at 14 30 21](https://github.com/shapely/shapely/assets/36797143/125f64fe-034f-49f4-9f11-b7257ef4c701)
Yet, it is present in the result.
![Screenshot 2024-03-25 at 14 31 21](https://github.com/shapely/shapely/assets/36797143/ec17f31f-4626-4053-98be-54b3e43274e6)

I would expect the result to not contain that small square. Is this my misunderstanding or a bug?

**Comment by theroggy** [comment]
I also think this is not the expected result, so I'd think it is a bug.

I had a (very) quick look on how it is implemented, and the same `Y_Y_reduce_func` is used as is used as for `intersection_all`. I didn't look into details how the c function works, but I suppose this implements a loop through all inputs where each input geometry is, for `symmetric difference_all`, `symmetric differenced` from the result of the previous two... which leads to this result.

**Comment by jorisvandenbossche** [comment]
Indeed, you essentially can't do this iteratively like we do now, in each step you need to consider all other geometries. 

The very dumb way I can currently think of is that for each geometry, you calculate the sym diff with the union (or just collection) of all other geometries. That sounds quite expensive, but not sure if there is a way to optimize this? (one could use a spatial index to do the sym diff with only the relevant (intersecting) other geometries)

@martinfleis did you have an actual use case for this, or just looking at it to consider adding the equivalent to geopandas? 
I am wondering if we shouldn't rather deprecate/remove this from shapely given it is not actually correct.

**Comment by jorisvandenbossche** [comment]
cc @caspervdw

**Comment by martinfleis** [comment]
> @martinfleis did you have an actual use case for this, or just looking at it to consider adding the equivalent to geopandas?
I am wondering if we shouldn't rather deprecate/remove this from shapely given it is not actually correct.

No use case. I hit this while trying to expose it on geopandas side. Deprecation is fine with me.

**Comment by paapu88** [comment]
I came across this with shapely and removing the middle box by intersection_all seems to work:
```
from shapely.geometry import box
from shapely import intersection_all, symmetric_difference_all
import numpy as np

geoms = np.array([box(0, 0, 2, 2), box(1, 1, 3, 3), box(0, 0, 1.5, 1.5)])
print(symmetric_difference_all(geoms))
print(symmetric_difference_all(geoms) - intersection_all(geoms))
```

**Comment by jorisvandenbossche** [comment]
Shall we deprecate this for shapely 2.1?

**Comment by theroggy** [comment]
> Shall we deprecate this for shapely 2.1?

I would. The current implementation is incorrect and fixing it takes an effort, and as there seems to be little animo on the need... Hence, better deprecate with the message that the current results can be incorrect... at least then it is/becomes clear to anyone that the current results cannot relied on at the moment.

**Comment by jorisvandenbossche** [comment]
Opened https://github.com/shapely/shapely/pull/2257 to do so

**Comment by Dhmss** [comment]
To me, the current behavior of `symmetric_difference_all ` is fine. It's a matter of specification, I understand that the current behavior may not be the expected one, but it is still entirely valid.

I use it to perform healing operations that are sensitive to contour winding orientation (clockwise or counter-clockwise), and I haven't found any alternative function in Shapely that can perform this operation.

What should I do if I need a function that preserves the current behavior of symmetric_difference_all ?

**Comment by Rupt** [comment]
I have swapped to `functools.reduce(shapely.symmetric_difference, geometries)` for my application of the deprecated `symmetric_difference_all`.

There may be users who enjoy the speed of having this reduction in the compiled library.
